slam-libraries/viam-cartographer/src/slam_service/config_test.cc

```diff
@@ -0,0 +1,338 @@
+    const std::string message = "slam mode is missing";
+    checkParseAndValidateConfigParamsException(argc, argv, message);
+    delete argv;
+}
+
+BOOST_AUTO_TEST_CASE(ParseAndValidateConfigParams_invalid_slam_mode) {
+    ResetFlagsForTesting();
+    std::vector<std::string> args{
+        "carto_grpc_server", "-config_param={mode=bad}", "-data_dir=/path/to",
+        "-port=localhost:0", "-sensors=lidar",           "-data_rate_ms=200",
+        "-map_rate_sec=60"};
+    int argc = args.size();
+    char** argv = toCharArrayArray(args);
+    const std::string message = "Invalid slam_mode=bad";
+    checkParseAndValidateConfigParamsException(argc, argv, message);
+    delete argv;
+}
+
+BOOST_AUTO_TEST_CASE(ParseAndValidateConfigParams_valid_config) {
+    ResetFlagsForTesting();
+    std::vector<std::string> args{
+        "carto_grpc_server", "-config_param={mode=2d}", "-data_dir=/path/to",
+        "-port=localhost:0", "-sensors=lidar",          "-data_rate_ms=200",
+        "-map_rate_sec=60"};
+    int argc = args.size();
+    char** argv = toCharArrayArray(args);
+    SLAMServiceImpl slamService;
+    config::ParseAndValidateConfigParams(argc, argv, slamService);
+
+    auto tolerance = boost::test_tools::tolerance(0.00001);
+    BOOST_TEST(slamService.path_to_data == "/path/to/data");
+    BOOST_TEST(slamService.path_to_map == "/path/to/map");
+    BOOST_TEST(slamService.config_params == "{mode=2d}");
+    BOOST_TEST(slamService.slam_mode == "2d");
+    BOOST_TEST(slamService.optimize_every_n_nodes == 3);
+    BOOST_TEST(slamService.num_range_data == 100);
+    BOOST_TEST(slamService.missing_data_ray_length == 25.0, tolerance);
+    BOOST_TEST(slamService.max_range == 25.0, tolerance);
+    BOOST_TEST(slamService.min_range == 0.2, tolerance);
+    BOOST_TEST(slamService.max_submaps_to_keep == 3);  // LOCALIZATION only
+    BOOST_TEST(slamService.fresh_submaps_count == 3);  // UPDATING only
+    BOOST_TEST(slamService.min_covered_area == 1.0,
+               tolerance);                                 // UPDATING only
+    BOOST_TEST(slamService.min_added_submaps_count == 1);  // UPDATING only
+    BOOST_TEST(slamService.occupied_space_weight == 20.0, tolerance);
+    BOOST_TEST(slamService.translation_weight == 10.0, tolerance);
+    BOOST_TEST(slamService.rotation_weight == 1.0, tolerance);
+    BOOST_TEST(slamService.port == "localhost:0");
+    BOOST_TEST(slamService.data_rate_ms.count() ==
+               std::chrono::milliseconds(200).count());
+    BOOST_TEST(slamService.map_rate_sec.count() ==
+               std::chrono::seconds(60).count());
+    BOOST_TEST(slamService.camera_name == "lidar");
+    BOOST_TEST(slamService.offlineFlag == false);
+    delete argv;
+}
+
+BOOST_AUTO_TEST_CASE(
+    ParseAndValidateConfigParams_valid_config_with_carto_params) {
+    ResetFlagsForTesting();
+    std::string config_param =
+        "{mode=2d,"
+        "optimize_every_n_nodes=5,num_range_data=500,"
```

Comment:
[nit] I think the test will be more likely to catch bugs if all of the parameters are set to different values.